util/network/src/node_table.rs

```diff
@@ -282,77 +286,63 @@ impl NodeTable {
 
 	/// Save the nodes.json file.
 	pub fn save(&self) {
-		if let Some(ref path) = self.path {
-			let mut path_buf = PathBuf::from(path);
-			if let Err(e) = fs::create_dir_all(path_buf.as_path()) {
-				warn!("Error creating node table directory: {:?}", e);
-				return;
-			};
-			path_buf.push("nodes.json");
-			let mut json = String::new();
-			json.push_str("{\n");
-			json.push_str("\"nodes\": [\n");
-			let node_ids = self.nodes(IpFilter::default());
-			for i in 0 .. node_ids.len() {
-				let node = self.nodes.get(&node_ids[i]).expect("self.nodes() only returns node IDs from self.nodes");
-				json.push_str(&format!("\t{{ \"url\": \"{}\", \"failures\": {} }}{}\n", node, node.failures, if i == node_ids.len() - 1 {""} else {","}))
-			}
-			json.push_str("]\n");
-			json.push_str("}");
-			let mut file = match fs::File::create(path_buf.as_path()) {
-				Ok(file) => file,
-				Err(e) => {
-					warn!("Error creating node table file: {:?}", e);
-					return;
+		let mut path = match self.path {
+			Some(ref path) => PathBuf::from(path),
+			None => return,
+		};
+		if let Err(e) = fs::create_dir_all(&path) {
+			warn!("Error creating node table directory: {:?}", e);
+			return;
+		}
+		path.push(NODES_FILE);
+		let node_ids = self.nodes(IpFilter::default());
+		let len = node_ids.len();
+		let nodes = node_ids.into_iter()
+			.map(|id| self.nodes.get(&id).expect("self.nodes() only returns node IDs from self.nodes"))
+			.map(|node| node.clone())
+			.filter(|node| if len > MAX_NODES { node.last_attempted.is_some() } else { true })
```

Comment:
Maybe we should change the nodes sorting so that nodes that we never connected to have a lower priority and then we `.take(MAX_NODES)` on the sorted list. This way we always get a node table with a predictable size.

Reply:
Sounds like a good idea, we might need to tune the comparator a bit though. Might be difficult to decide what nodes are more valuable: never connected or connected once, but with a single failure.
In general I need to double check if we reset `failures` counter in case we successfuly connect to the node, otherwise valuable peers might be dropped to often (with time they would accumulate a lot of failures, so the preference would be to completely new nodes).